File: leetcode2044/solution.go
/*
Solution: backtracking

Time: O(2^n) | Space: O(n)

Runtime: 5 ms | 86.21%
Memory: 3.88 MB | 21.74%

- n: length of 'nums'
*/

func countMaxOrSubsets(nums []int) int {
    mx := 0
    for _, num := range nums {
        mx |= num
    }

    res := 0;

    backtrack(&res, 0, mx, 0, nums);

    return res;
}

func backtrack(res *int, path int, mx int, current int, nums []int) {
    if current == len(nums) {
        if path == mx {
            *res += 1;
        }
        return;
    } else {
        backtrack(res, path, mx, current + 1, nums);
        backtrack(res, path | nums[current], mx, current + 1, nums);
    }
}

File: leetcode2044/solution.rs
/*
Solution: backtracking

Time: O(2^n) | Space: O(n)

Runtime: 0 ms | 100.00%
Memory: 2.10 MB | 38.46%

- n: length of 'nums'
*/

impl Solution {
    pub fn count_max_or_subsets(nums: Vec<i32>) -> i32 {
        let mx: i32 = nums.iter().fold(0, |acc, &num| acc | num);

        let mut res: i32 = 0;

        Self::backtrack(&mut res, 0, mx, 0, &nums);

        res
    }

    fn backtrack(res: &mut i32, path: i32, mx: i32, current: usize, nums: &Vec<i32>) {
        if current == nums.len() {
            *res += if path == mx {1} else {0};
            return;
        }

        Self::backtrack(res, path, mx, current + 1, nums);
        Self::backtrack(res, path | nums[current], mx, current + 1, nums);
    }
}